File: solutions/javascript/linked-list/1/linked-list.js
//
// This is only a SKELETON file for the 'Linked List' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class LinkedList {
  head = undefined;
  tail = undefined;
  length = 0;

  //end of list
  push(value) {
    this.length++;
    const newNode = new Node(value);

    if (!this.tail || !this.head) {
      this.tail = newNode;
      this.head = newNode;
      return;
    }

    this.tail.next = newNode;
    newNode.prev = this.tail;
    this.tail = newNode;
    return;
  }

  //end of list
  pop() {
    if (!this.tail) return;

    this.length--;

    const deadNode = this.tail;

    if (this.tail.prev) {
      this.tail.prev.next = undefined;
    }

    this.tail = this.tail.prev;

    deadNode.prev = undefined;

    return deadNode.value;
  }

  //remove from beginning
  shift() {
    if (!this.head) return;

    this.length--;

    const deadNode = this.head;
    if (this.head.next) {
      this.head.next.prev = undefined;
    }

    this.head = this.head.next;

    deadNode.next = undefined;

    return deadNode.value;
  }

  //add to beginning
  unshift(value) {
    this.length++;
    const newNode = new Node(value);

    if (!this.head || !this.tail) {
      this.head = newNode;
      this.tail = newNode;
      return;
    }

    this.head.prev = newNode;
    newNode.next = this.head;

    this.head = newNode;
  }

  delete(value) {
    let current = this.head;

    while (current) {
      if (current.value === value) {
        this.length--;

        if (this.length === 0) {
          this.head = undefined;
          this.tail = undefined;
          return;
        }

        if (current === this.head) {
          this.head = current.next;
          if (this.head) {
            this.head.prev = undefined;
          }

          return;
        }

        if (current === this.tail) {
          this.tail = current.prev;

          if (this.tail) {
            this.tail.next = undefined;
          }

          return;
        }

        current.prev.next = current.next;
        current.next.prev = current.prev;

        return;
      }

      current = current.next;
    }
  }

  count() {
    return this.length;
  }
}

class Node {
  value;
  next;
  prev;

  constructor(value) {
    this.value = value;
    this.next = undefined;
    this.prev = undefined;
  }
}

File: solutions/ruby/linked-list/1/linked_list.rb
# Write your code for the 'Linked List' exercise in this file. Make the tests in
# `linked_list_test.rb` pass.
#
# To get started with TDD, see the `README.md` file in your
# `ruby/linked-list` directory.
class Node
  attr_accessor :value, :next, :prev

  def initialize(value)
    @value = value
    @next = nil
    @prev = nil
  end
end

class Deque
  def initialize
    @head = nil
    @tail = nil
    @length = 0
  end

  def push(value)
    new_node = Node.new(value)

    unless @tail.nil?
      @tail.next = new_node
      new_node.prev = @tail
    end

    @head = new_node if @head.nil?

    @length += 1

    @tail = new_node
  end

  def pop
    unless @tail.nil?
      dead_node = @tail
      @tail = @tail.prev

      @length -= 1

      @head = nil if @length.zero?

      return dead_node.value
    end

    nil
  end

  def shift
    unless @head.nil?
      dead_node = @head
      @head = @head.next
      @length -= 1

      @tail = nil if @length.zero?

      return dead_node.value
    end

    nil
  end

  def unshift(value)
    new_node = Node.new(value)

    unless @head.nil?
      @head.prev = new_node
      new_node.next = @head
    end

    @head = new_node

    @length += 1

    @tail = new_node if @tail.nil?
  end
end
